shapely/set_operations.py

```diff
@@ -501,3 +503,110 @@ def coverage_union_all(geometries, axis=None, **kwargs):
     # create_collection acts on the inner axis
     collections = lib.create_collection(geometries, GeometryType.GEOMETRYCOLLECTION)
     return lib.coverage_union(collections, **kwargs)
+
+
+@multithreading_enabled
+def disjoint_subset_union(a, b, **kwargs):
+    """Merges multiple polygons into one. This is an optimized version of
+    union which assumes inputs can be divided into subsets that do not intersect.
+
+    If there is only one such subset, performance can be expected to be worse than
+    :func:`union`. As such, it is recommeded to use ``disjoint_subset_union`` with
+    GeometryCollections rather than individual geometries.
+
+    Requires GEOS >= 3.12.0
+
+    Parameters
+    ----------
+    a : Geometry or array_like
+    b : Geometry or array_like
+    **kwargs
+        See :ref:`NumPy ufunc docs <ufuncs.kwargs>` for other keyword arguments.
+
+    See also
+    --------
+    union
+    coverage_union
+    disjoint_subset_union_all
+
+    Examples
+    --------
+    >>> from shapely import normalize, Polygon
+    >>> polygon_1 = Polygon([(0, 0), (0, 1), (1, 1), (1, 0), (0, 0)])
+    >>> polygon_2 = Polygon([(1, 0), (1, 1), (2, 1), (2, 0), (1, 0)])
+    >>> normalize(disjoint_subset_union(polygon_1, polygon_2))
+    <POLYGON ((0 0, 0 1, 1 1, 2 1, 2 0, 1 0, 0 0))>
+
+    Union with None returns same polygon:
+
+    >>> normalize(disjoint_subset_union(polygon_1, None))
+    <POLYGON ((0 0, 0 1, 1 1, 1 0, 0 0))>
+    """
+    if (isinstance(a, Geometry) or a is None) and (
+        isinstance(b, Geometry) or b is None
+    ):
+        pass
+    elif isinstance(a, Geometry) or a is None:
+        a = np.full_like(b, a)
+    elif isinstance(b, Geometry) or b is None:
+        b = np.full_like(a, b)
+    elif len(a) != len(b):
+        raise ValueError("Arrays a and b must have the same length")
+    return disjoint_subset_union_all([a, b], axis=0, **kwargs)
+
+
+@multithreading_enabled
+def disjoint_subset_union_all(geometries, axis=None, **kwargs):
+    """Returns the union of multiple polygons of a geometry collection.
```

Comment:
```suggestion
    """Returns the union of multiple polygons.
```

The "of a geometry collection" might be a bit confusing, given we create the collection under the hood?